### DISCUSSION TITLE
experiments code, or example train and generation?

### DISCUSSION THREAD

**Discussion by rlabs-oss** [archinetai/audio-diffusion-pytorch#8] [General]
An example showing instantiation from a checkpoint to generate audio would be appreciated.

**Comment by flavioschneider** [comment]
You are passing a dataset to the input instead of a batch of elements, please have a look at the examples in the readme to understand the correct tensor shape.

**Reply by ModeratePrawn** [reply]
I think that some people finding this repo and wanting to use it to experiment with diffusion audio generation are used to using things like Stable Diffusion and Dalle that have scripts available to generate the images without having to know how to code in python, or have deep understanding of the models and how they work. It might be useful to provide some example python scripts like the colab notebook that people can run locally to easily train a model and then use the model (or a pre-trained checkpoint) to generate some audio. (Like the reference sample script here: https://github.com/CompVis/stable-diffusion/blob/main/scripts/txt2img.py). That may be easier than a step-by-step guide, though some would no doubt find that useful as well.

**Reply by flavioschneider** [reply]
There's this [colab notebook](https://colab.research.google.com/gist/flavioschneider/39c6454bfc2d03dc7d0c5c9d8e2078d1/audio-diffusion-pytorch-v0-1.ipynb) to test the functionality on a pretrained model. Other than that, at the moment this model very experimental and the trainer is meant to be used by researchers interested in exploring different architectures or settings. It will likely change a lot in the future, and when it's working properly, I'm sure that I or someone else will create a way to train the model more easily.

**Reply by rlabs-oss** [reply]
Trying the notebook in a local / none-colab environment.  What am I missing here? 

RuntimeError                              Traceback (most recent call last)
Input In [7], in <cell line: 10>()
      5 model = AudioDiffusionModel(in_channels=2, diffusion_sigma_data=0.2)
      7 #state_dict = torch.load('epoch2512-valid_loss0052.ckpt')  #load checkpoint from file test
      8 #model.load_state_dict(state_dict)
      9 ##online_dict = torch.hub.load_state_dict_from_url('https://cdn-lfs.huggingface.co/repos/69/d8/69d850779d5624398eda53943e1868592bac47ef66b2e25352489dd777e610f9/19ee5eee902f552514de301d110d28a81847e1f8cec3a04baf0494c668fa1187')
---> 10 model.load_state_dict(torch.hub.load_state_dict_from_url('https://cdn-lfs.huggingface.co/repos/69/d8/69d850779d5624398eda53943e1868592bac47ef66b2e25352489dd777e610f9/19ee5eee902f552514de301d110d28a81847e1f8cec3a04baf0494c668fa1187'))

File ~/audiovenv2/lib/python3.8/site-packages/torch/nn/modules/module.py:1604, in Module.load_state_dict(self, state_dict, strict)
   1599         error_msgs.insert(
   1600             0, 'Missing key(s) in state_dict: {}. '.format(
   1601                 ', '.join('"{}"'.format(k) for k in missing_keys)))
   1603 if len(error_msgs) > 0:
-> 1604     raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
   1605                        self.__class__.__name__, "\n\t".join(error_msgs)))
   1606 return _IncompatibleKeys(missing_keys, unexpected_keys)

RuntimeError: Error(s) in loading state_dict for AudioDiffusionModel:
	size mismatch for unet.downsamples.0.downsample.weight: copying a param with shape torch.Size([256, 4, 9]) from checkpoint, the shape in current model is torch.Size([256, 128, 9]).
	size mismatch for unet.downsamples.1.downsample.weight: copying a param with shape torch.Size([512, 4, 9]) from checkpoint, the shape in current model is torch.Size([512, 256, 9]).

**Reply by rlabs-oss** [reply]
This is a discussion under discussions in how to research and fault find the model/checkpoints and other environmental issues that may crop up.

**Reply by flavioschneider** [reply]
Have you installed version `0.0.15`?

**Reply by rlabs-oss** [reply]
0.0.19 was installed, reverted back to 0.0.15 and restarted the kernel - the checkpoint via url returns "All keys matched successfully"

Thank you

**Reply by matigekunstintelligentie** [reply]
When I change the model to a checkpoint, which was resumed from the pretrained model, I trained with the [audio-diffusion-pytorch-trainer](https://github.com/archinetai/audio-diffusion-pytorch-trainer) repo. I get missing keys too. I only changed 
```model.load_state_dict(torch.load("last.ckpt"))```
The yaml has in_channels=2, diffusion_sigma_data=0.2
I've also tried both 0.0.19 and 0.0.15 to no avail.
What am I doing wrong?

**Reply by flavioschneider** [reply]
The created checkpoint is a pytorch-lightning `.ckpt` checkpoint, not directly compatible with pytorch `.pt` checkpoints. Please check the FAQ in the README.

**Reply by rlabs-oss** [reply]
In practice that's something resembling:- 
The current directory is the trainer folder (audio_diffusion_pytorch_trainer)

```
from main.module_base import Model

model = Model.load_from_checkpoint(
    checkpoint_path='/logs/ckpts/YYYY-MM-DD-HH-MM-SS/last.ckpt',
    learning_rate=1e-4,
    beta1=0.9,
    beta2=0.99,
    in_channels=2,
    patch_size=16
)
torch.save(model.model, 'torchckpt.pt')
```

With instantiation using one of the existing code examples;
```

model=torch.load('torchckpt.pt', map_location="cuda")

with torch.no_grad():
.... rest of code...

```

Play from notebook, or save sample:
```
mysample = Audio(sample_long.squeeze().cpu(), rate = sampling_rate)

with open('generated_test.wav', 'wb') as f:
    f.write(mysample.data)

```

**Comment by rlabs-oss** [comment]
above example no longer works with the latest version. 

Im attempting (with no examples)

from main.module_base import Model
#main.module_base.Model

model = Model.load_from_checkpoint(
    checkpoint_path='logs/ckpts/2022-10-10-23-42-25/last.ckpt',
    lr=1e-4,
    lr_beta1=0.95,
    lr_beta2=0.999,
    lr_eps=1e-6,
    lr_weight_decay=1e-3,
    ema_beta=0.9999,
    ema_power=0.7,
    model=AudioDiffusionModel  << not sure thats correct?
)

which also doesnt work.

   1878         Module: self
   1879     """
-> 1880     for p in self.parameters():
   1881         p.requires_grad_(requires_grad)
   1882     return self

AttributeError: 'bool' object has no attribute 'parameters'

parameters obfuscated in more parameters.
